src/Symfony/Bundle/SecurityBundle/Security/FirewallMap.php

```diff
@@ -141,15 +139,27 @@ public function getFirewallConfig(Request $request)
         return $context->getConfig();
     }
 
+    /**
+     * @param Request $request
+     *
+     * @return \Symfony\Bundle\SecurityBundle\Security\FirewallContext
+     */
     private function getFirewallContext(Request $request)
     {
-        if ($this->contexts->contains($request)) {
-            return $this->contexts[$request];
+        if ($request->attributes->has('_firewall_context')) {
+            $contextId = $request->attributes->get('_firewall_context');
+            if (array_key_exists($contextId, $this->map)) {
```

Comment:
`$this->map` can be `Traversable` (and [it is](https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php#L255))